omnigibson/sensors/vision_sensor.py

```diff
@@ -215,38 +208,75 @@ def initialize_sensors(self, names):
 
         # Run super first to grab any upstream obs
-        obs = super()._get_obs()
+        obs, info = super()._get_obs()
+
+        for modality in self._modalities:
+            raw_obs = self._annotators[modality].get_data()
+            # Obs is either a dictionary of {"data":, ..., "info": ...} or a direct array
+            obs[modality] = raw_obs["data"] if isinstance(raw_obs, dict) else raw_obs
+            if modality == "seg_semantic":
+                id_to_labels = raw_obs['info']['idToLabels']
+                obs[modality], corrected_id_to_labels = self._remap_semantic_segmentation(obs[modality], id_to_labels)
+                info[modality] = corrected_id_to_labels
+            elif modality == "seg_instance":
+                id_to_labels = raw_obs['info']['idToLabels']
+                info[modality] = id_to_labels
+        return obs, info
+    
+    def _remap_semantic_segmentation(self, img, id_to_labels):
+        """
+        Remap the semantic segmentation image to the class IDs defined in CLASS_NAME_TO_CLASS_ID.
+        Also, correct the id_to_labels input with the labels from CLASS_NAME_TO_CLASS_ID and return it.
+        """
+        # Convert string IDs to integers
+        int_ids = set(int(id) for id in id_to_labels.keys())
+
+        # Determine if there are any new IDs that were not previously known
+        new_ids = int_ids - self._known_semantic_ids
+
+        # If there are new IDs, update _known_semantic_ids set and rebuild the key array
+        if new_ids:
+            self._known_semantic_ids.update(new_ids)
+            max_id = max(self._known_semantic_ids)
+
+            # Initialize the key array with a default value for unmapped IDs
+            key_array = np.full(max_id + 1, -1, dtype=np.int32)
+
+            # Populate the key array with the new IDs based on class name mappings
+            for int_id in self._known_semantic_ids:
+                str_id = str(int_id)
+                if str_id in id_to_labels:
+                    info = id_to_labels[str_id]
+                    class_name = info['class'].lower()
+                    if class_name in CLASS_NAME_TO_CLASS_ID:
+                        new_id = CLASS_NAME_TO_CLASS_ID[class_name]
+                        key_array[int_id] = new_id
+        else:
+            # Use the existing key_array if no new IDs are found
+            key_array = self.key_array
 
-        # Process each sensor modality individually
-        for modality in self.modalities:
-            mod_kwargs = dict()
-            mod_kwargs["viewport"] = self._viewport.viewport_api
-            if modality == "seg_instance":
-                mod_kwargs.update({"parsed": True, "return_mapping": False})
-            elif modality == "bbox_3d":
-                mod_kwargs.update({"parsed": True, "return_corners": True})
-            obs[modality] = self._SENSOR_HELPERS[modality](**mod_kwargs)
+        corrected_id_to_labels = {}
```

Comment:
shouldn't this also be stateful -- i.e.: it gets aggregated over time?

Reply:
Here's my understanding of this: replicator returns 1) the segmentation map 2) id-to-labels mapping for objects we are currently looking at. This remap function returns 1) the remapped segmentation map 2) the remapped id-to-labels for the subset of objects we are currently looking at. The second returned map is essentially `corrected_id_to_labels`. By stateful/aggregated over time, do you mean that this should include global information (i.e. return the class variable `known_semantic_ids`) instead of the local subset of id mapping?